### ISSUE TITLE
Feature: ONNX / TensorRT export for trained policies

### ISSUE THREAD

**Issue by jashshah999** [huggingface/lerobot#3146]
### Ticket Type

Feature Request / Improvement

### Description

LeRobot covers the full pipeline from data collection to training to async inference, but there's currently no built-in way to export a trained policy (ACT, Diffusion Policy, etc.) to optimized formats like ONNX or TensorRT for edge deployment.

This is a common pain point for anyone deploying on resource-constrained hardware (Jetson Orin Nano, etc.) where PyTorch inference is too slow or memory-heavy. Right now the only option is D-Robotics' hardware-specific [rdk_LeRobot_tools](https://github.com/D-Robotics/rdk_LeRobot_tools), but there's nothing general-purpose.

### Proposed scope

A `lerobot-export` CLI command (or equivalent API) that:

1. **Loads a trained policy checkpoint** (local path or Hub repo)
2. **Exports to ONNX** with correct dynamic axes for batch size / sequence length
3. **Optionally optimizes with TensorRT** (FP16/INT8) when available
4. **Validates correctness** by comparing PyTorch vs exported model outputs on sample inputs

Example usage:
```bash
lerobot-export \
  --policy.path=lerobot/act_pusht \
  --format=onnx \
  --output=act_pusht.onnx

# With TensorRT optimization
lerobot-export \
  --policy.path=lerobot/act_pusht \
  --format=tensorrt \
  --precision=fp16 \
  --output=act_pusht.engine
```

### Why this matters

- The SO-101 + Jetson Orin Nano combo is increasingly popular and PyTorch inference can be a bottleneck
- ONNX export enables deployment across diverse hardware (Jetson, Intel NPUs, Qualcomm, etc.) without vendor lock-in
- TensorRT typically gives 2-5x inference speedup over vanilla PyTorch on NVIDIA GPUs
- This would complement the existing async inference stack by making the policy server itself faster

### What I've looked into

- `torch.onnx.export` works for ACT and Diffusion Policy architectures with some attention to dynamic axes
- The main challenge is handling policy-specific preprocessing/postprocessing (action chunking, denoising loops for diffusion, etc.)
- VLA models (Pi0, SmolVLA) are harder due to size but ONNX export is still feasible; TensorRT would require more work

Happy to put together a PR if this is something the team would want upstream. Would also appreciate guidance on which policies to prioritize.

**Comment by rylinjames** [comment]
this is exactly the gap i kept hitting deploying act + diffusion to jetson. fwiw `torch.onnx.export` on act works but the tricky part is action chunking and the diffusion denoising loop, you basically have to either export the unet alone and keep the scheduler in python, or trace the full k-step loop with a fixed horizon (loses flexibility but way faster on trt). on orin nano / orin nx you want fp16 onnx + trt, not int8 or fp8 (sm_87 has no fp8 tensor cores, trt will silently fall back or fail build). been working on reflex-vla, open-source deploy layer for VLAs that does exactly this pipeline (onnx export + fp16 trt + cos-sim parity check) for smolvla/pi0/pi0.5/gr00t. github.com/rylinjames/reflex-vla. we tagged good-first-issues (including one for an so-100 example) if you want to poke around, or dm romirj on discord / playindus@gmail.com for collaborator access.

**Comment by rylinjames** [comment]
hey, built this already -- it's called reflex-vla (github.com/FastCrest/reflex-vla)

covers smolvla, pi0, pi0.5, and gr00t. monolithic onnx export with verified cos=+1.0 parity against pytorch, trt-ep baked in. the smolvla export was genuinely annoying to get right (broken kv cache wiring, wrong sinusoidal embedding, image dim mismatch) -- took a bunch of patches to torch.export to get there.

would be useful to hear if you hit anything it doesn't cover for your use case

**Comment by tsuu-abj** [comment]
My apologies—I completely missed reflex-vla! Thanks for bringing it to my attention. Achieving +1.0 parity for smolvla with those torch.export patches is a serious achievement; I’ll definitely be studying your implementation closely.

For my focus on real-world robot deployment, having this level of optimized export natively in LeRobot would be a game-changer. Would you be open to contributing these improvements as a PR to LeRobot? It would be a massive win for the ecosystem to have your work available as a standard feature.

That said, I realize there might be significant technical hurdles or architectural reasons for keeping reflex-vla separate—my apologies if I’m overlooking those complexities.

**Comment by rylinjames** [comment]
thanks -- yeah the torch.export patches were a rabbit hole. pi0 alone had three interacting bugs that only showed up when you ran the full denoising loop.

contributing pieces to lerobot makes sense eventually but the runtime side (TRT EP, edge serving, the wedge system) doesn't really belong in a training framework. the exporter could be a PR but it'd be the least interesting part without the rest.

genuinely curious what you're deploying on -- policy on the robot or offloading to a workstation?

**Comment by SeungWooPark0507** [comment]
Hi, thanks for opening this issue. I’m also very interested in edge deployment of LeRobot policies.

Could you share a bit more about the current status of the ONNX work?

In particular, I’m curious about:

- how far the ONNX export and validation work has progressed so far,
- which policy architecture is currently considered the highest-priority target (ACT, Diffusion Policy, SmolVLA, etc.),
- whether there is already a working branch/prototype for any of them,
- and whether the longer-term scope is primarily model inference, or an end-to-end deployable pipeline including policy-specific preprocessing and postprocessing.

I’m especially interested in edge inference, so understanding the first target policy and the intended deployment boundary would be very helpful.
